File: win-prog/ProgWinCS/Beziers and Other Splines/ClosedCurveFillModes/ClosedCurveFillModes.cs
//---------------------------------------------------
// ClosedCurveFillModes.cs © 2001 by Charles Petzold
//---------------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

class ClosedCurveFillModes: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new ClosedCurveFillModes());
     }
     ClosedCurveFillModes()
     {
          Text = "FillClosedCurve Fill Modes";
          ClientSize = new Size(2 * ClientSize.Height, ClientSize.Height);
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          Brush   brush = new SolidBrush(clr);
          Point[] apt   = new Point[5];
          
          for (int i = 0; i < apt.Length; i++)
          {
               double dAngle = (i * 0.8 - 0.5) * Math.PI;
               apt[i] = new Point(
                              (int)(cx *(0.25 + 0.24 * Math.Cos(dAngle))),
                              (int)(cy *(0.50 + 0.48 * Math.Sin(dAngle))));
          }
          grfx.FillClosedCurve(brush, apt, FillMode.Alternate);

          for (int i = 0; i < apt.Length; i++) 
               apt[i].X += cx / 2;         

          grfx.FillClosedCurve(brush, apt, FillMode.Winding);
     }
}

File: win-prog/ProgWinVB/Beziers and Other Splines/ClosedCurveFillModes/ClosedCurveFillModes.vb
'-----------------------------------------------------
' ClosedCurveFillModes.vb (c) 2002 by Charles Petzold
'-----------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Class ClosedCurveFillModes
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(New ClosedCurveFillModes())
    End Sub

    Private Sub New()
        Text = "FillClosedCurve Fill Modes"
        ClientSize = New Size(2 * ClientSize.Height, ClientSize.Height)
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Dim br As New SolidBrush(clr)
        Dim apt(4) As Point
        Dim i As Integer

        For i = 0 To apt.GetUpperBound(0)
            Dim rAngle As Double = (i * 0.8 - 0.5) * Math.PI
            apt(i) = New Point( _
                        CInt(cx * (0.25 + 0.24 * Math.Cos(rAngle))), _
                        CInt(cy * (0.5 + 0.48 * Math.Sin(rAngle))))
        Next i

        grfx.FillClosedCurve(br, apt, FillMode.Alternate)

        For i = 0 To apt.GetUpperBound(0)
            apt(i).X += cx \ 2
        Next i

        grfx.FillClosedCurve(br, apt, FillMode.Winding)
    End Sub
End Class
